### PR TITLE
k-point symmetry for mean-field methods and mp2

### PR DESCRIPTION
**Pull request by fishjojo**

### CODE REVIEW

**Comment by sunqm** [comment]
Looks good to me.  Just one concern, does the code in pbc/tools/k2gamma.py that transforms k-point sampling to super gamma point compatible with k-point symmetry?

**Comment by zhcui** [comment]
Do we have a plan to merge this recently? I think it would be useful.

**Comment by Exferro** [comment]
Hi everyone!
Do I understand it correctly that this is the branch providing the "experimental support for additional lattice symmetries" as stated in ["Recent developments in the PySCF program package"](https://arxiv.org/pdf/2002.12531.pdf)?

If so, could you please tell how safe would it be to use it — I saw it was planned to be included in the latest release, but am not sure if it goes ahead.

And also, do I understand it right that support of additional lattice symmetries means obtaining symmetry-adapted crystalline orbitals similarly to what happens in molecular systems when we account for the point group symmetries and obtain symmetry-adapted molecular orbitals?

Thank you!

**Comment by fishjojo** [comment]
Hi @Exferro, currently, this only provides reduced k points in the Brillouin zone. Symmetry adapted orbitals will be added in the near future.

**Review by sunqm — APPROVED** [review]

File: pyscf/pbc/lib/kpts_helper.py [hunk 1]
```diff
@@ -32,6 +32,27 @@ def is_zero(kpt):
...
+        wrap_around : bool
+            If set to True, k-points are rounded to [-0.5, 0.5), otherwise [0.0, 1.0).
+            Default value is False.
+        tol : float
+            K-points differ less than tol are considered as the same.
+            Default value is 1e-6.
+    '''
+    kpts_round = np.mod(kpts, 1)
+    kpts_round = kpts_round.round(-np.log10(tol).astype(int))
+    kpts_round = np.mod(kpts_round, 1)
```
[the remark this answers is not in the document]
Comment (fishjojo) [reply pyscf/pbc/lib/kpts_helper.py:51]: I vaguely remember there were some edge cases that required this. But I could be wrong.

**Comment by verena-neufeld** [comment]
Hi Xing, I just wanted to check that the use of `unique` here is stable when floating point numbers are involved.

**Comment by fishjojo** [comment]
> Hi Xing, I just wanted to check that the use of `unique` here is stable when floating point numbers are involved.

The numbers should have been rounded to certain precision before the operation of np.unique, so it should be fine. There may be some edge cases where it doesn't work, and we can fix them later.

**Comment by tberkel** [comment]
For context, we recently found an issue with unique kpts affecting
`get_monkhorst_pack_size()`
https://github.com/pyscf/pyscf/blob/0a6569604b9133665708bfe28d9491aa453cb4a5/pyscf/pbc/tools/pbc.py#L467
and thus the Ewald energy for certain kpt meshes. PR coming soon.

**Comment by verena-neufeld** [comment]
As an example, we had the problem that due to numerical noise, two scaled k point coordinates that should be the same number, were `0.05167749999999998` and `0.05167750000000004`, being rounded to `0.051677` and `0.051678` respectively and therefore found not to be the same.

**Comment by sunqm** [comment]
The numerical error in kpts does not caused by the k-point symmetry treatments. It seems a universal problem for the error tolerance in k points. The tolerance are handled in various places, like the unique funciton in kpts_helper, the kpts_to_kmesh function, get_monkhorst_pack_size etc. I think this PR just needs to consider perfect symmetry in k space. The error tolerance can be studied in a separated pull request.

**Pull request merged**